src/orange/detail/BIHIntersectingVolFinder.hh

```diff
@@ -0,0 +1,260 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2022-2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file orange/detail/BIHIntersectingVolFinder.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "corecel/math/Algorithms.hh"
+
+#include "BIHView.hh"
+#include "../BoundingBoxUtils.hh"
+#include "../OrangeData.hh"
+#include "../univ/detail/Types.hh"
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Traverse the BIH to the find the volume that the ray intersects with first.
+ *
+ * Traversal is carried out using a depth first search. During traversal, the
+ * minimum intersection is stored.  The decision to traverse an edge is done by
+ * calculating the distance to intersection with the precomputed edge bounding
+ * box. The edge bounding box is the bounding box created by clipping an
+ * infinite bounding box with all bounding planes between the root node and the
+ * current edge (inclusive). If a ray's intersection with the edge bbox is
+ * found to be nearer than the current minimum intersection, traversal procedes
+ * down that edge. Likewise, when a root node is reacted, intersections with
+ * volume bboxes are first tested against the minimum intersection prior to
+ * testing the the volume itself. The minimum intersection is only modified
+ * when a nearer minimumium intersection with a actual volume if found, NOT a
+ * nearer intersection with an edge bbox or volume bbox. This is because is is
+ * possible to have a ray that interects with a volume's bbox, but not the
+ * volume itself.
+ *
+ * \todo move to top-level orange directory out of detail namespace
+ */
+class BIHIntersectingVolFinder
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using Storage = NativeCRef<BIHTreeData>;
+
+    struct Ray
+    {
+        Real3 pos;
+        Real3 dir;
+    };
+    //!@}
+
+    // Construct from a vector of bounding boxes and storage for LocalVolumeIds
+    inline CELER_FUNCTION
+    BIHIntersectingVolFinder(BIHTree const& tree, Storage const& storage);
+
+    // Calculate the minimum intersection
+    template<class F>
+    inline CELER_FUNCTION Intersection operator()(Ray const& ray,
+                                                  F&& visit_vol) const;
+
+  private:
+    //// DATA ////
+    BIHView view_;
+
+    //// HELPER FUNCTIONS ////
+
+    // Get the ID of the next node in the traversal sequence
+    inline CELER_FUNCTION BIHNodeId next_node(BIHNodeId const& current_id,
+                                              BIHNodeId const& previous_id,
```

Comment:
Since NodeID are POD , it's better to pass by value (and that way the compiler can save some work trying to figure out if the IDs are aliased or not).